core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -259,12 +259,15 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
   def prepareProducerIdRotation(newProducerId: Long,
                                 newTxnTimeoutMs: Int,
                                 updateTimestamp: Long,
-                                recordLastEpoch: Boolean): TxnTransitMetadata = {
-    if (hasPendingTransaction)
+                                recordLastEpoch: Boolean,
+                                fenceProducer: Boolean): TxnTransitMetadata = {
+    if (hasPendingTransaction && !fenceProducer)
       throw new IllegalStateException("Cannot rotate producer ids while a transaction is still pending")
 
-    prepareTransitionTo(Empty, newProducerId, 0, if (recordLastEpoch) producerEpoch else RecordBatch.NO_PRODUCER_EPOCH,
-      newTxnTimeoutMs, immutable.Set.empty[TopicPartition], -1, updateTimestamp)
+    logger.info(s"Rotating producer ID from $producerId to $newProducerId because the producer epoch was exhausted")
+    prepareTransitionTo(if (fenceProducer) PrepareEpochFence else Empty, newProducerId, 0,
```

Comment:
Similarly, let's pass a new state instead of a boolean which needs to be determined again.

Reply:
The problem I have with that is that there are only two states we should ever transition to in this method. A caller shouldn't be allowed to pass, say, `CommitAbort`. Keeping the logic in here means that the caller only has to know whether it's a fencing rotation or a normal rotation.